photosynthesis_metrics/perceptual.py

```diff
@@ -0,0 +1,280 @@
+"""
+Implementation of VGG16 loss, originaly used for style transfer and usefull in many other task (including GAN training)
+"""
+from typing import List, Union
+
+import torch
+import torch.nn as nn
+from torch.nn.modules.loss import _Loss
+from torchvision.models import vgg16, vgg19
+
+from photosynthesis_metrics.utils import _validate_input, _adjust_dimensions
+
+
+# Map VGG names to corresponding number in torchvision layer
+VGG16_LAYERS = {
+    "conv1_1": 0, "relu1_1": 1,
+    "conv1_2": 2, "relu1_2": 3,
+    "pool1": 4,
+    "conv2_1": 5, "relu2_1": 6,
+    "conv2_2": 7, "relu2_2": 8,
+    "pool2": 9,
+    "conv3_1": 10, "relu3_1": 11,
+    "conv3_2": 12, "relu3_2": 13,
+    "conv3_3": 14, "relu3_3": 15,
+    "pool3": 16,
+    "conv4_1": 17, "relu4_1": 18,
+    "conv4_2": 19, "relu4_2": 20,
+    "conv4_3": 21, "relu4_3": 22,
+    "pool4": 23,
+    "conv5_1": 24, "relu5_1": 25,
+    "conv5_2": 26, "relu5_2": 27,
+    "conv5_3": 28, "relu5_3": 29,
+    "pool5": 30,
+}
+
+VGG19_LAYERS = {
+    "conv1_1": 0, "relu1_1": 1,
+    "conv1_2": 2, "relu1_2": 3,
+    "pool1": 4,
+    "conv2_1": 5, "relu2_1": 6,
+    "conv2_2": 7, "relu2_2": 8,
+    "pool2": 9,
+    "conv3_1": 10, "relu3_1": 11,
+    "conv3_2": 12, "relu3_2": 13,
+    "conv3_3": 14, "relu3_3": 15,
+    "conv3_4": 16, "relu3_4": 17,
+    "pool3": 18,
+    "conv4_1": 19, "relu4_1": 20,
+    "conv4_2": 21, "relu4_2": 22,
+    "conv4_3": 23, "relu4_3": 24,
+    "conv4_4": 25, "relu4_4": 26,
+    "pool4": 27,
+    "conv5_1": 28, "relu5_1": 29,
+    "conv5_2": 30, "relu5_2": 31,
+    "conv5_3": 32, "relu5_3": 33,
+    "conv5_4": 34, "relu5_4": 35,
+    "pool5": 36,
+}
+
+
+class ContentLoss(_Loss):
+    r"""Creates Content loss that can be used for image style transfer of as a measure in
+    image to image tasks.
+    Uses pretrained VGG models from torchvision. Normalizes features before summation.
+    Expects input to be in range [0, 1] or normalized with ImageNet statistics into range [-1, 1]
```

Comment:
Range [-1, 1] is invalid according to checks in `_validate_input`.
Same for following losses.

Reply:
Btw it is interesting why tests did not fail because of that. I think it shows that tests simply do not cover this case, which has to be fixed